[DCS Patch 2/5] port dynamic channel selection from rdkb infrastructure

itayx opened the pull request:
> The Dynamic Channel Selection as a feature can provide information on the supported channels which can be utilized by the management system to provide better availability and user experience.
#562

The Dynamic Channel Selection feature is divided into several pull request, this is the second out of 5
start review from 9db41a9


### High Level Infastructure (Controller & yaml): Support for Channel Scan & DCS:
Changes in this PR include the following:
- beerocks defines: enums for Channel Scan & DCS
- yaml: beerocks_message_common: Add channel scan structures
- yaml: beerocks_message_control: Add channel scan & DCS CMDUs
- controller: node: Add channel scan structures
- controller: db: Add channel scan DB methods

itayx committed:
beerocks defines: enums for Channel Scan & DCS

This is a preperative commit for 'Add channel scan structures'.

Add enums to support the DCS feature.
Add eChannelScanErrCode that describes the scan operation status.
Add eDcsOpErrCode that describes the API's operation status.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: node: Add channel scan structures

This is a preperative commit for 'Add channel scan DB methods'.

Add the result class & container, which is used to store
the DCS scan results returned from the monitor/agent.
Add DCS configutation class for the DCS feature.
Add DCS scan status class, which is used to
store the latest DCS scan results status.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
controller: db: Add channel scan DB methods

Add DB method get_hostap_by_mac
Add DB methods to get/set scan params (configutation).
Add DB methods to get/set scan status.
Add DB methods to get/set scan results.
Add DB methods to clear channel pool.
Add DB methods to check if a channel is in the
current channel scan's pool.

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -443,6 +446,30 @@ enum eBssType {
 };
 
+enum eChannelScanErrCode : uint8_t {
+    CHANNEL_SCAN_NO_ERROR = 0,
```
Why not `CHANNEL_SCAN_SUCCESS`?
Also, why not use enum class instead of enum?
```
enum class eChannelScanErrCode : uint8_t { SUCCESS, INTERNAL_FAILURE, TRIGGERED_EVENT_TIMEOUT, ... };

itayx replied:
changed it to enum class and changed "NO_ERROR" to "SUCCESS"

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -443,6 +446,30 @@ enum eBssType {
+    CHANNEL_SCAN_INVALID_PARAMS
+};
+
+enum eChannelScanOpErrCode : uint8_t {
```
same here - use enum class when possible.

tomereli commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_action.h:
```diff
@@ -103,7 +103,13 @@ enum eActionOp_CONTROL: uint8_t {
     ACTION_CONTROL_STEERING_EVENT_PROBE_REQ_NOTIFICATION = 0x85,
     ACTION_CONTROL_STEERING_EVENT_AUTH_FAIL_NOTIFICATION = 0x86,
-    ACTION_CONTROL_ENUM_END = 0x87,
+    ACTION_CONTROL_CHANNEL_SCAN_TRIGGER_SCAN_REQUEST = 0x8c,
```
Please split this PR to 2 commits - the one that is adding the YAMLs, and the one that includes the autogenerated files.

itayx replied:
split the commits between the YAML & AutoGenerated

tomereli commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_bml.yaml:
```diff
@@ -362,3 +362,82 @@ cACTION_BML_TRIGGER_CHANNEL_SELECTION_REQUEST:
+  _type: class
+  op_error_code:
+    _type: uint8_t
+    _comment: 0 - No Error (Success), Otherwise error according to beerocks_defines:eDcsOpErrCode
```
`_comment: 0 - Success, Otherwise error according to ...`

itayx replied:
Fixed

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -33,6 +33,28 @@ typedef struct {
 } sVapElement;
 
+typedef struct {
+    // std::string radio;
```
Please remove unused comment

tomereli replied:
Please add doxygen for new structs and comment on each member

itayx replied:
removed the whole structure using the one in beerocks_message_common

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -33,6 +33,28 @@ typedef struct {
+typedef struct {
+    // std::string radio;
+    std::string ssid;
+    std::string mac;
```
Please us `sMacAddr` for mac addresses where possible. no need to store strings in the db if not absolutely needed.

itayx replied:
removed the whole structure using the one in beerocks_message_common (which uses sMacAddr)

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -33,6 +33,28 @@ typedef struct {
+    // std::string radio;
+    std::string ssid;
+    std::string mac;
+    std::string mode;
```
feels like mode should be an enumeration. Please explain each parameter if it is a string then why.

itayx replied:
removed the whole structure using the one in beerocks_message_common
which uses enums (also in beerocks_message_common)

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +205,27 @@ class node {
         std::shared_ptr<ap_stats_params> stats_info;
         std::unordered_map<int8_t, sVapElement> vaps_info;
+
+        class channel_scan_config {
```
If its a class without methods and everything public - then why not declare it as struct?

itayx replied:
changed to struct

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +205,27 @@ class node {
+            int dwell_time_msec = -1;       //-1 (invalid)
+        };
+
+        class channel_scan_status {
```
Same here - should be struct.

itayx replied:
changed to struct

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +205,27 @@ class node {
+            bool scan_in_progress                              = false;
+            beerocks::eChannelScanErrCode last_scan_error_code = beerocks::CHANNEL_SCAN_NO_ERROR;
+        };
+        std::shared_ptr<channel_scan_config> continuous_scan_config;
```
why is everything `shared_ptr` here? Why not just regular members? Feels like a waste for no good reason.

vitalybu replied:
I agree. Result in too many memory allocations for no apparent reason.

itayx replied:
We followed other examples in the code, and they did the same structure

> class X {
> ...
> };
> shared_ptr<X> x;;

tomereli replied:
Please change that. No need for shared_ptr.

itayx replied:
changed

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+//
+// Channel Scan
+//
+bool db::set_channel_scan_is_enabled(const std::string &mac, const bool enable)
```
I think it would be better to split to 2 functions:
`channel_scan_enable()` and `channel_scan_disable()` (nitpick)

itayx replied:
this is basically a set for a flag, since we don't do any additional logical manipulations, there is no need to split the function

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+        LOG(ERROR) << "node not found.... ";
+        return false;
+    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
+        LOG(ERROR) << __FUNCTION__ << "node " << mac << " is not a valid hostap!";
```
perhaps print the type also in the error log can be useful for debug.

itayx replied:
Added a print for the type

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+        LOG(ERROR) << __FUNCTION__ << "node " << mac << " is not a valid hostap!";
+        return false;
+    }
+    LOG(DEBUG) << __FUNCTION__ << ", continuous scan is enable = " << int(enable);
```
I prefer to get rid of the "is_enable" and just change it to "enable".
So:
`LOG(DEBUG) << "continues scan " << enable ? "enabled" : "disabled";` will be much better.

tomereli replied:
#665

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+    return n->hostap->continuous_scan_config->is_enabled;
+}
+
+bool db::set_channel_scan_interval_sec(const std::string &mac, const int interval_sec)
```
Can `interval_sec` be negative? If not, please change to `unsigned int`.

vitalybu replied:
Also, no need to add `const` for primitive types

itayx replied:
while it can't be negative, the default is set to -1, in order to force the user to initially set the `interval_sec`, either we declare a default for the `interval_sec` or have the member and function signature not match

tomereli replied:
so just remove the `const`

itayx replied:
removed all the unneeded `const` from the signatures

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+    }
+
+    //Validate new dcs channel pool with hostap_supported_channels list
+    //This is a workaround
```
Not OK - the supported channels should not be declared staticly but received from the agent.
@YuriyMasechko @arnout please comment.

kantera800 replied:
This will be fixed as a separate task (already have one in the intel backlog)
start using GET_HW_FEATURES for channel scan pool validation.
I'll open task in github.

kantera800 replied:
opened #686

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+    return scan_config->channel_pool;
+}
+
+bool db::is_channel_in_pool(const std::string &mac, const uint8_t channel, const bool single_scan)
```
no need to declare simple types as `const` - see CONTRIBUTING.md.

itayx replied:
removed unneeded const

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+
+bool db::add_channel_scan_results(const std::string &mac,
+                                  const sChannelScanResultsElement &scan_result,
+                                  const bool single_scan)
```
`bool single_scan`. Same comment for all other places in this file.

itayx replied:
removed unneeded const

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+        return dummy_return;
+    }
+
+    auto &scan_results =
```
no need for the assignment if you're going to return it right after.
`return (single_scan) ? n->hostap->single_scan_results : n->hostap->continuous_scan_results;`

tomereli replied:
#665

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+                                                                          const bool single_scan)
+{
+    auto n = get_node(mac);
+    static std::list<sChannelScanResultsElement> dummy_return;
```
Oh this one is super ugly.
I prefer `return std::list<sChannelScanResultsElement>();` instead everywhere.
Or you can return a shared_ptr instead of const reference and return nullptr instead of dummy_return.

Anything BUT `dummy_return` 👎

vitalybu replied:
`return {}` can work as well

itayx replied:
Tried that, when I do I get an error saying:
> error: returning reference to temporary [-Werror=return-local-addr]

arnout replied:
> Or you can return a shared_ptr instead of const reference and return nullptr instead of dummy_return.

This will be the only thing that works. But it's super-annoying, because it means the caller has to check for nullptr everywhere while it could just work with the empty set without problem.

tomereli replied:
> > Or you can return a shared_ptr instead of const reference and return nullptr instead of dummy_return.
> 
> This will be the only thing that works. But it's super-annoying, because it means the caller has to check for nullptr everywhere while it could just work with the empty set without problem.

So you think the `dummy_return` should stay?

itayx replied:
for now it stays, if we decide to fix this later on we'll decide in #665

tomereli commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -394,6 +394,42 @@ class db {
+    // Channel Scan
+    //
+    bool set_channel_scan_is_enabled(const std::string &mac, const bool enable);
+    bool get_channel_scan_is_enabled(const std::string &mac);
```
here and in any other method which doesn't change the state of the class, declare the function as const:
```
bool get_channel_scan_is_enabled(const std::string &mac) const;
```

tomereli replied:
Also - the name is really long and for no reason. I prefer `channel_scan_enabled(const std::string &mac)` (nitpick)

itayx replied:
So it will be
> channel_scan_enabled(const std::string &mac) const; //get
> channel_scan_enabled(const std::string &mac, const bool enable); //set

?

tomereli replied:
There are far more important things to fix in this PR. So let's keep it for #665

tomereli commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -394,6 +394,42 @@ class db {
 
+    //
+    // Channel Scan
+    //
```
Please add Doxygen comments to all functions here or in the implementation file.

itayx replied:
Added

tomereli commented on controller/src/beerocks/master/db/db.h:
```diff
@@ -394,6 +394,42 @@ class db {
+    //
+    // Channel Scan
+    //
+    bool set_channel_scan_is_enabled(const std::string &mac, const bool enable);
```
I prefer `bool channel_scan_enable(const std::string &mac, bool enable);`
Same for all other APIs.
(nitpick)

itayx replied:
most other function in the db follow the set_X get_X signature style

tomereli commented on common/beerocks/tlvf/AutoGenerated/include/beerocks/tlvf/beerocks_message_common.h:
```diff
@@ -1189,6 +1189,92 @@ typedef struct sSteeringEvSnr {
+    }
+} __attribute__((packed)) sChannelScanRequestParams;
+
+typedef struct sChannelScanResults {
```
I think we talked about this one - it should be refactored to not using strings, so it should first be done here.

itayx replied:
refactored

vitalybu commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -59,7 +59,7 @@ if(BWL_TYPE STREQUAL "DWPAL")
 
     find_package(dwpal REQUIRED)
-    list(APPEND BWL_LIBS dwpal)
+    list(APPEND BWL_LIBS dwpal nl-3 nl-genl-3)
```
See comment from previous PR

tomereli replied:
Fixed in #680

vitalybu commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -100,7 +100,14 @@ eActionOp_CONTROL:
+  ACTION_CONTROL_CHANNEL_SCAN_ABORT_NOTIFICATION : 144
+  ACTION_CONTROL_CHANNEL_SCAN_FINISHED_NOTIFICATION : 145
+
+  ACTION_CONTROL_ENUM_END: 150
```
Why not 146?

itayx replied:
changed

tomereli replied:
I think it make sense to leave "space" for future actions. this will make it easier to add new ones without rewriting the end enum.

arnout replied:
> I think it make sense to leave "space" for future actions. this will make it easier to add new ones without rewriting the end enum.

In fact, the _END values are not used anywhere. Why do we add them?

vitalybu commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:
```diff
@@ -282,7 +289,16 @@ eActionOp_MONITOR:
+  ACTION_MONITOR_CHANNEL_SCAN_ABORT_NOTIFICATION: 66
+  ACTION_MONITOR_CHANNEL_SCAN_FINISHED_NOTIFICATION: 67
+
+  ACTION_MONITOR_ENUM_END: 70
```
68?

itayx replied:
changed

vitalybu commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -33,6 +33,28 @@ typedef struct {
+    std::string operating_frequency_band;
+    std::string supported_standards;
+    std::string operating_standards;
+    std::string operating_channel_bandwidth;
```
As discussed yesterday, all string should be changed to simple types

itayx replied:
removed the whole structure, using the one in beerocks_message_common (which uses simple types and enums also defined in beerocks_message_common)

tomereli replied:
@itayx please refer to the PR which fixes this in this PR description.

vitalybu commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +205,27 @@ class node {
+        class channel_scan_config {
+        public:
+            bool is_enabled = false;
+            std::set<uint8_t> channel_pool; // default value: empty list
```
Why is this a `set` and not `vector` or even an `unordered_set`?

itayx replied:
since we run the std::set::find function in the db::is_channel_in_pool function
other then that no reason

tomereli replied:
http://www.cplusplus.com/reference/unordered_set/unordered_set/find/

itayx replied:
We used std::set since the find is most efficient in an ordered set
but since the list is short it's negligible so..
changing to unordered_set

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+        return false;
+    }
+    return n->hostap->continuous_scan_config->is_enabled;
+}
```
This `if` statement repeats in almost every function.
Please create a local static function that returns a pointer to the `hostap` structure, and simply return `nullptr` on error.
It'll greatly reduce the code base.

For example:
```
static <hostap*> get_hostap_by_mac(const std::string &mac) {
    auto n = get_node(mac);

    if (!n) {
        LOG(ERROR) << "node not found.... ";
        return nullptr;
    } else if (n->get_type() != beerocks::TYPE_SLAVE || n->hostap == nullptr) {
        LOG(ERROR) << "node " << mac << " is not a valid hostap!";
        return nullptr;
    }

    return n->hostap;
}
```

itayx replied:
Added a private member function
> std::shared_ptr\<node::radio\> get_hostap_by_mac(const std::string& mac);

in both db.h & .cpp

Could not use a static function since `get_node` is also a private member function

vitalybu commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+                       << " is invalid, setting channel pool failed !";
+            return false;
+        }
+    }
```
What's the purpose of this test?
It seems like you build a list of ALL available channels, and just check that the given channel is in this list.
Do we expect here just a valid channel or we need to make sure that the Agent actually supports this channel?

itayx replied:
When setting the channel pool, we need to make sure that all the channels in the new pool are supported, we tried using existing lists in the db to compare to, but we couldn't find any that did so we defaulted to a static list of the supported channels

tomereli replied:
The supported channels should come from BWL - and not declared as const. #665

kantera800 replied:
opened separate task for this 
see https://github.com/prplfoundation/prplMesh/pull/643#discussion_r367404728

vitalybu commented on common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:
```diff
@@ -826,4 +826,101 @@ sSteeringEvSnr:
+  operating_standards:
+    _comment: Comma-separated list of strings. Each list item MUST be a member of the list reported by theSupportedStandardsparameter. List items indicate which IEEE 802.11 standard that is detected for thisResult.
+    _type: char
+    _length: ["beerocks::message::WIFI_OPERATING_STRING_LENGTH"]
```
Just a reminder to drop all the strings in favor of simpler types...

itayx replied:
refactored to use enums 👍

arnout commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -2117,6 +2117,324 @@ bool db::get_hostap_is_acs_enabled(std::string mac)
+    auto n = get_node(mac);
+    if (!n) {
+        LOG(ERROR) << "node not found.... ";
+        return *(std::make_shared<std::set<uint8_t>>());
```
Does this work? AFAICS this has the same problem as `return std::set<uint8_t>();` only the compiler doesn't complain about it...

tomereli replied:
This is basically creating a shared_ptr, then instead of returning the shared_ptr itself which would increment the reference count preventing the memory from being released, we return the contents and when we return they are invalid.

itayx replied:
Added a dummy_return, if we decide that using a dummy_return is incorrect we'll fix it in #665

tomereli reviewed (commented):
one small comment

kantera800 commented:
ready for re-review:
1. fixed all code review issues  that not require push -f .
2. clang format ran only from icsl.

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -33,27 +33,43 @@ typedef struct {
+//     uint32_t channel_utilization;
+//     //Timestamp for the channel results
+//     std::chrono::steady_clock::time_point timestamp;
+// } sChannelScanResultsElement;
```
Why not delete the comments?

itayx replied:
deleted

tomereli reviewed (commented):
Please rebase-squash and I'll do the followup review.
@morantr please review after the rebase.

itayx committed:
TLVF: beerocks_message_common: Add channel scan structures

This is a preperative commit for 'Add channel scan & DCS CMDUs'.

Add DCS structures to support the DCS feature [yaml only].

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
TLVF: beerocks_message_common: channel scan structure autogenerated

This is a preperative commit for 'Add channel scan & DCS CMDUs'.

Add DCS structures to support the DCS feature [AutoGenerated].

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
TLVF: beerocks_message_control: Add channel scan & DCS CMDUs

Add requests from the bml to the controller [YAML].
Add requests from the controller to the agent [YAML].
Add notifications from the agent to the controller [YAML].
Add requests from the agent to the monitor [YAML].
Add notifications from the monitor to the agent [YAML].

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

itayx committed:
TLVF: beerocks_message_control: channel scan & DCS CMDUs autogenerated

Add requests from the bml to the controller [AutoGenerated].
Add requests from the controller to the agent [AutoGenerated].
Add notifications from the agent to the controller [AutoGenerated].
Add requests from the agent to the monitor [AutoGenerated].
Add notifications from the monitor to the agent [AutoGenerated].

Signed-off-by: itay elenzweig <itayx.elenzweig@intel.com>

kantera800 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [[DCS Patch 2/5]Port dynamic channel selection](https://trello.com/c/ZyRbblaJ/76-dcs-patch-2-5port-dynamic-channel-selection)

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -443,6 +447,30 @@ enum eBssType {
 };
 
+enum class eChannelScanErrCode : uint8_t {
+    CHANNEL_SCAN_SUCCESS = 0,
```
Its an enum class - no need for the `CHANNEL_SCAN_` prefix. #665

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_defines.h:
```diff
@@ -443,6 +447,30 @@ enum eBssType {
+};
+
+enum class eChannelScanOpErrCode : uint8_t {
+    CHANNEL_SCAN_OP_SUCCESS = 0,
```
Here as well

kantera800 replied:
#665

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +184,28 @@ class node {
+                beerocks::eChannelScanErrCode::CHANNEL_SCAN_SUCCESS;
+        };
+        //These members are part of the continuous scan
+        //The contiuous scan, scans every
```
The continuos scan, scan every ? every what?

kantera800 replied:
added interval_sec

tomereli commented on controller/src/beerocks/master/db/node.h:
```diff
@@ -183,6 +184,28 @@ class node {
+        };
+        //These members are part of the continuous scan
+        //The contiuous scan, scans every
+        channel_scan_config continuous_scan_config;
```
Please add doxygen for each new parameter, what it is and what it is used for.

kantera800 replied:
done

tomereli commented on controller/src/beerocks/master/db/db.cpp:
```diff
@@ -3137,6 +3386,22 @@ std::shared_ptr<node> db::get_node(sMacAddr al_mac, sMacAddr ruid)
     return get_node(key);
 }
 
+std::shared_ptr<node::radio> db::get_hostap_by_mac(const std::string &mac)
```
Please change to
`std::shared_ptr<node::radio> db::get_hostap_by_mac(const sMacAddr &mac)`

No new code will be merged which does all the string to mac and mac to string conversions now that we have all that we need as part of the sMacAddr structure.

kantera800 replied:
done

tomereli reviewed (approved):
Small comments - but please fix them before merging. Especially everything related to the string to mac and mac to string conversions.
